From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Thibaud Michaud <thibaudm@chromium.org>
Date: Tue, 17 Sep 2024 16:49:30 +0200
Subject: [M126-LTS][wasm] Check strict type equality for Tag imports

R=manoskouk@chromium.org

(cherry picked from commit f612d9a40b194cc7fc1d9cffdd295a78fa2f10c9)

Fixed: 365802567
Change-Id: I38d70f157f9a78fe56eb0c377776dfe794872473
Reviewed-on: https://chromium-review.googlesource.com/c/v8/v8/+/5868875
Commit-Queue: Thibaud Michaud <thibaudm@chromium.org>
Reviewed-by: Manos Koukoutos <manoskouk@chromium.org>
Cr-Original-Commit-Position: refs/heads/main@{#96143}
Reviewed-on: https://chromium-review.googlesource.com/c/v8/v8/+/5872320
Reviewed-by: Thibaud Michaud <thibaudm@chromium.org>
Commit-Queue: Gyuyoung Kim (xWF) <qkim@google.com>
Cr-Commit-Position: refs/branch-heads/12.6@{#64}
Cr-Branched-From: 3c9fa12db3183a6f4ea53d2675adb66ea1194529-refs/heads/12.6.228@{#2}
Cr-Branched-From: 981bb15ba4dbf9e2381dfc94ec2c4af0b9c6a0b6-refs/heads/main@{#93835}

diff --git a/src/wasm/wasm-objects.cc b/src/wasm/wasm-objects.cc
index c0c340d253370d8c443323577e9695975c899d28..7cf7fdef1c38b07cfc487cd0ebad70e1fae94bf3 100644
--- a/src/wasm/wasm-objects.cc
+++ b/src/wasm/wasm-objects.cc
@@ -1813,8 +1813,8 @@ Handle<WasmTagObject> WasmTagObject::New(Isolate* isolate,
 }
 
 bool WasmTagObject::MatchesSignature(uint32_t expected_canonical_type_index) {
-  return wasm::GetWasmEngine()->type_canonicalizer()->IsCanonicalSubtype(
-      this->canonical_type_index(), expected_canonical_type_index);
+  return static_cast<uint32_t>(this->canonical_type_index()) ==
+         expected_canonical_type_index;
 }
 
 const wasm::FunctionSig* WasmCapiFunction::GetSignature(Zone* zone) const {
